### PR TITLE
Src 6613 update controller guis

### PR DESCRIPTION
**Pull request by georgiablanco**
## Proposed changes

Updating change controller gui to include a Direct PWM commands and a stop all control options:
image
image
image

image

adding more information: 
image

## Checklist

Before posting a PR ensure that from each of the below categories **AT LEAST ONE BOX HAS BEEN CHECKED**. If more than one category is applicable then more can be checked. Also ensure that the proposed changes have been filled out with relevant information for reviewers.

## Tests

- [ ] No tests required to be added. (For small changes that will be tested by CI/CD infrastructure).
- [ ] Added automated tests (if a new class is added (Python or C++), the interface of that class must be unit tested).
- [ ] Manually tested in simulation (if simulation specific or no hardware required to test the functionality). 
- [x] Manually tested on hardware (if hardware specific or related).

## Documentation

- [ ] No documentation required to be added.
- [x] Added documentation (For any new feature, explain what it does and how to use it. Write the documentation in a relevant space, e.g. Github, Confluence, etc).
- [ ] Updated documentation (For changes to pre-existing features mentioned in the documentation).

### CODE REVIEW

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 1]
```diff
@@ -135,7 +156,15 @@ def confirm_current_control(self):
             rospy.logerr(error + ". Error: " + err)
             return
 
-        running_controllers = [c for c in resp1.controller if c.state == "running"]
+        # Confirm all running controllers for the controls in this GUI
+        running_controllers = [c for c in resp1.controller
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: Could we use a better name than `c`?

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 2]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+                            if all(hand in self._controller_groups for hand in ["rh_", "lh_"]):
+                                hand_id = 'sr_bimanual_hands_robot'
+                            else:
+                                hand_id = 'sr_hand_robot'
+                            srv_path = '/{}/{}/change_control_type'.format(hand_id, robot_name[:2])
+                            query_control_type = rospy.ServiceProxy(srv_path, ChangeControlType)
+                            try:
+                                change_type_msg = ChangeControlType()
+                                change_type_msg.control_type = ControlType.QUERY
+                                resp1 = query_control_type(change_type_msg)
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: Perhaps something more descriptive than `resp1` here?

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 3]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+
+                                if resp1.result.control_type == 0:
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]:
```suggestion
                                if resp1.result.control_type == ControlType.PWM:
```

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 4]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+                                    current_robot_control[robot_name] = 3
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: `= 3` isn't very informative. You could do something like `current_robot_control[robot_name] = RobotTeachModeRequest.DIRECT_PWM_MODE` instead, although it's not really a teach mode request here. Or, you could maybe define some constants instead?

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 5]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+                                elif resp1.result.control_type == 1:
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: same as above

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 6]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+                                    current_robot_control[robot_name] = 2
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: same as above

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 7]
```diff
@@ -153,12 +182,34 @@ def confirm_current_control(self):
...
+                            except rospy.ServiceException as err:
+                                error = "Couldn't get control type from {} service".format(srv_path)
+                                QMessageBox.warning(self._widget, "No Control Type Found", error)
+                                rospy.logerr(error + ". Error: " + err)
+                                return
+                        else:
+                            current_robot_control[robot_name] = 2
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: and here

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 8]
```diff
@@ -260,14 +314,45 @@ def display_information(self, message):
                   "The output of the host side PID controller is sent to the motor " + \
                   "as a PWM demand. No effort controller is used for position control.\n" + \
                   "Teach Mode: No control is implemented on the host. The Effort " + \
-                  "demand is sent to the motor which implements it using a 5kHz control loop."
+                  "demand is sent to the motor which implements it using a 5kHz control loop.\n" + \
+                  "Direct PWM Commands: This is used for basic position control, " + \
+                  "and is used by default on a new hand. The PWM demand value is sent " + \
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: what do you mean by "used by default on a new hand"?

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 9]
```diff
@@ -260,14 +314,45 @@ def display_information(self, message):
...
+            'controller_manager/switch_controller', SwitchController)
+
+        controllers_to_start = []
+        controllers_to_stop = []
+        for controller in running_controllers:
+            if robot_prefix in controller.name:
+                controllers_to_stop.append(controller.name)
+
+        try:
+            resp1 = switch_controllers(
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: `resp1` again

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py [hunk 10]
```diff
@@ -276,7 +361,7 @@ def change_teach_mode(mode, robot):
         req = RobotTeachModeRequest()
         req.teach_mode = mode
         req.robot = robot
-        modes_str = ["TRAJECTORY_MODE", "TEACH_MODE", "POSITION_MODE", "GRASP_MODE"]
+        modes_str = ["TRAJECTORY_MODE", "TEACH_MODE", "POSITION_MODE", "DIRECT_PWM_MODE"]
```
Comment (carebare47) [inline sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py:367]: How come this is defined twice?

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 1 above)
Comment (carebare47) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: actually, maybe `resp1` too

**Comment by carebare47** [comment]
In testing, if teach mode is requested via service call before starting this plugin, the green squares don't line up with the currently running controller until after 'stop all controllers' has been pressed

Upon starting the plugin (with teachmode already running)
Screenshot from 2022-01-12 17-58-19

When changing to direct pwm with this plugin:
Screenshot from 2022-01-12 17-58-57

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 8 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: i found this in the palm_EDC_user_manual.pdf
image

Comment (carebare47) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: I think that's referring to the types of command that can be sent to the motor, where PWM demand is a direct motor instruction (used with position control, trajectory control, direct PWM/effort), and Torque demand instead uses a Torque control loop on the motor board (used for teach mode only). So, PWM demand is default for a new hand _in the provided launch files_, but that doesn't extend to 'Direct PWM Commands' in change_controllers.py

Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: do you think just removing that part sounds better:
`"Direct PWM Commands: This is used for basic position control. The PWM demand value is sent directly to the motor, unless there is a safety cutout.\n"`

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 7 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 6 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 5 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 4 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 3 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 2 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 1 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed in [542d322](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/542d322c257524753a6d7f478c22fe42ffc70d69)

File: sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py (hunk 8 above)
Comment (georgiablanco) [reply sr_gui_change_controllers/src/sr_gui_change_controllers/change_controllers.py]: fixed [9932816](https://github.com/shadow-robot/sr_visualization_common/pull/35/commits/99328168fdda7403a372d20431ce791cec249269)

**Review by carebare47 — APPROVED** [review]

**Review by biofotis — APPROVED** [review]

**Pull request merged**